unittests/runtime/Concurrent.cpp

```diff
@@ -42,3 +44,48 @@ TEST(ConcurrentReadableArrayTest, SingleThreaded) {
   add(1000000);
   check();
 }
+
+TEST(ConcurrentReadableArrayTest, MultiThreaded) {
+  const int writerCount = 16;
+  const int readerCount = 8;
+  const int insertCount = 100000;
+
+  struct Value {
+    int threadNumber;
+    int x;
+  };
+  ConcurrentReadableArray<Value> array;
+
+  auto writer = [&](int threadNumber) {
+    for (int i = 0; i < insertCount; i++)
+      array.push_back({ threadNumber, i });
+  };
+
+  auto reader = [&] {
+    int maxByThread[writerCount];
+    bool done = false;
+    while (!done) {
+      for (int i = 0; i < writerCount; i++)
+        maxByThread[i] = -1;
+      for (auto element : array.snapshot()) {
+        ASSERT_LT(element.threadNumber, writerCount);
+        ASSERT_GT(element.x, maxByThread[element.threadNumber]);
```

Comment:
Comment might be helpful here:  It took me a minute to decipher that you're verifying that entries added by each writer are in increasing order.